### DISCUSSION TITLE
imgfetch read-only RAMDISK

### DISCUSSION THREAD

**Discussion by kingsumos** [ipxe/ipxe#649] [Q&A]
When booting Shell.efi, iPXE creates a RAMDISK containing all the downloaded files.
This filesystem, shown as FS0:, is a read-only filesystem. It is possible to make this filesystem writable? I have some EFI applications requiring write access...

iPXE script:

#!ipxe
dhcp net0
imgfetch http://${next-server}/Shell.efi
imgfetch http://${next-server}/startup.nsh
boot

UEFI Shell output:

UEFI Interactive Shell v2.2
EDK II
UEFI v2.70 (EDK II, 0x00010000)
Mapping table
      FS0: Alias(s):F0:;BLK0:
          PciRoot(0x0)/Pci(0x17,0x0)/Pci(0x0,0x0)/MAC(001AE8C36D70,0x1)
Press ESC in 3 seconds to skip startup.nsh or any other key to continue.
FS0:\> dir
Directory of: FS0:\
00/00/0000  00:00       r       3,468  startup.nsh
00/00/0000  00:00       r     951,744  Shell.efi
          2 File(s)     955,212 bytes
          0 Dir(s)
FS0:\>
FS0:\> edit test.txt
edit: Access Denied
FS0:\>

**Comment by NiKiZe** [comment]
Have you tried to see if you can enable any kind of iPXE debugging that can provide info on where this fails?

It might be possible.

**Reply by kingsumos** [reply]
The boot works, the UEFI Shell is loaded and we have a command prompt (as shown above in the example output).
But the RAM filesystem containing the files is READONLY, perhaps something is missing in iPXE.efi to make the filesystem writable. Or, the RAM filesystem was set by iPXE to read-only on purpose.
IMHO, debugging will not help in this case since iPXE already done it's job... it's Shell.efi executing/loaded inside a read-only RAM filesystem.

**Reply by NiKiZe** [reply]
You get access denied, that does not say anything about the FS. My assumption is that iPXE returns this, and I call that access denied a failure (non success return to the call)

You call this a read-only RAM filesystem, but there is nothing that indicates that efi is aware about it not being writable. But when being attempted it fails. I would rather call this write protected media.

**Comment by stappersg** [comment]
@kingsumos express what you further want and what it is worth to you.

**Reply by kingsumos** [reply]
Hi guys, 

I have already tried to enable DEBUG logs in iPXE to see if something is shown in the console output but with no luck (it's is a headless server with a serial console), nothing is shown in the log when trying to write a file. IMHO, iPXE has already passed the control to UEFI Shell and all the interactions are then between UEFI Shell and the UEFI FW.

However, I'm an EDK2 platform integrator of the UEFI FW running in this server - I have already enabled the DEBUG logs in UEFI and I can tell you that it's not a problem of any missing UEFI PROTOCOL, it looks like this RAM FW was really set to readonly - still missing more tests to confirm this, maybe I'll need to use the HW debugger (Lauterbach) since for those lower layers we don't have too much DEBUG lines in UEFI FW/EDK2.

Meanwhile, would be great if you can provide more information of how iPXE creates the block device were the FS0 filesystem is stored:
FS0: Alias(s):F0:;BLK0:
PciRoot(0x0)/Pci(0x17,0x0)/Pci(0x0,0x0)/MAC(001AE8C36D70,0x1)

I have an additional question, there is any way to reuse the network configuration from the initial boot? I mean, UEFI PXE FW downloads the iPXE.efi (from 'next-server' address) application then executes it. Once iPXE.efi is running, it needs to configure the network again (via DHCP or static IP's) for downloading the other files, e.g. kernel, initrd (in my case shell.efi, startup.ns h and others). The network configuration takes a while, it would be nice if we have access to the initial network configuration (done by UEFI PXE) and reuse it.

Kind regards,
Sumo

**Reply by stappersg** [reply]
This comment might not be wise.  Wise is
> Do not tell it can't be done to those who are doing it

And it is not wise to ignore (new) team members.

Now here the actual comment:

As I see it:
* _iPXE_ is just a bootloader
* bootloaders need RAM to store what they fetch
* _UEFI_ provides RAM
* _UEFI_ provides abstraction layer  _"FS0:"_
* _FS0:_ was never intented to be a full-fledged  filesystem
* Starting point _"seeing some files in something that looks a file systems, means being close to a writeable filesystem"_ is way too optimistic
* Observation _"it looks like this RAM FW was really set to readonly"_ sounds correct   (Please doubt me)
* It will be a long journey (Please be warned)


Thing that the journey already is bringing is https://www.markdownguide.org/extended-syntax/#fenced-code-blocks
so

FS0: Alias(s):F0:;BLK0:
PciRoot(0x0)/Pci(0x17,0x0)/Pci(0x0,0x0)/MAC(001AE8C36D70,0x1)

versus nice formatted
```text
FS0: Alias(s):F0:;BLK0:
PciRoot(0x0)/Pci(0x17,0x0)/Pci(0x0,0x0)/MAC(001AE8C36D70,0x1)
```


_Bon voyage_
